src/KeyRegistry.sol

```diff
@@ -0,0 +1,299 @@
+     */
+    function bulkAddKeysForMigration(
+        uint256[] calldata fids,
+        bytes[][] calldata fidKeys,
+        bytes calldata metadata
+    ) external onlyOwner {
+        if (isMigrated() && block.timestamp > keysMigratedAt + gracePeriod) revert Unauthorized();
+        if (fids.length != fidKeys.length) revert InvalidBatchInput();
+
+        // Safety: i and j can be incremented unchecked since they are bound by fids.length and
+        // fidKeys[i].length respectively.
+        unchecked {
+            for (uint256 i = 0; i < fids.length; i++) {
+                uint256 fid = fids[i];
+                for (uint256 j = 0; j < fidKeys[i].length; j++) {
+                    // TODO: add note about griefing during migration
+                    _add(fid, 1, fidKeys[i][j], metadata);
+                }
+            }
+        }
+    }
+
+    /**
+     * @notice Reset multiple keys as part of the initial migration. Only callable by the contract owner.
+     *         Reset is not the same as removal: this function sets the key state back to NULL,
+     *         rather than REMOVED. This allows the owner to correct any errors in the initial migration until
+     *         the grace period expires.
+     *
+     * @param fids    A list of fids whose added keys should be removed.
+     * @param fidKeys A list of keys to remove for each fid, in the same order as the fids array.
+     */
+    function bulkResetKeysForMigration(uint256[] calldata fids, bytes[][] calldata fidKeys) external onlyOwner {
+        if (isMigrated() && block.timestamp > keysMigratedAt + gracePeriod) revert Unauthorized();
+        if (fids.length != fidKeys.length) revert InvalidBatchInput();
+
+        // Safety: i and j can be incremented unchecked since they are bound by fids.length and
+        // fidKeys[i].length respectively.
+        unchecked {
+            for (uint256 i = 0; i < fids.length; i++) {
+                uint256 fid = fids[i];
+                for (uint256 j = 0; j < fidKeys[i].length; j++) {
+                    // TODO: add note about griefing during migration
+                    _reset(fid, fidKeys[i][j]);
+                }
+            }
+        }
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                                 HELPERS
+    //////////////////////////////////////////////////////////////*/
+
+    function _add(uint256 fid, uint32 scheme, bytes calldata key, bytes calldata metadata) internal {
+        KeyData storage keyData = keys[fid][key];
+        if (keyData.state != KeyState.NULL) revert InvalidState();
+
+        keyData.state = KeyState.ADDED;
+        keyData.scheme = scheme;
+        emit Add(fid, scheme, key, key, metadata);
+    }
+
+    function _reset(uint256 fid, bytes calldata key) internal {
+        KeyData storage keyData = keys[fid][key];
+        if (keyData.state != KeyState.ADDED) revert InvalidState();
```

Comment:
Why have this check? Could be more useful to be able to clear the map regardless of state.

Reply:
I went back and forth on this. Rationale is that this will help us prevent migration errors where we might incorrectly seed an array. I don't hold this strongly, and we might change our minds on this after doing the testnet migration.